### ISSUE TITLE
Switch Language

### ISSUE THREAD

**Issue by tyras2** [cyberjunk/meridian59-dotnet#118]
Would be fine if there is an solution, to load the german language files.

**Comment by cyberjunk** [comment]
All languages are already loaded from the RSB file and can be accessed, but
currently two things are still missing for full multilanguage support:

1) Received ServerStrings can be processed in english only, due to missing support for the new parameter-reordering which occurs in them for anything but english.

2) The static strings in the UI are hardcoded to english right now. OgreClient needs a proper way to switch all the internal strings between languages.

**Comment by Ketzer2002** [comment]
I'll try my best to work on this Issue

**Comment by Ketzer2002** [comment]
@cyberjunk Is there a possibility to talk in german on a channel to this issue? I have one or two questions. :-)

**Comment by Ketzer2002** [comment]
At first glance the static strings in the UI mentioned under 2) could be edited in this file:

"meridian59-dotnet/Resources/ui/layouts/Meridian59.layout"

Is there a possibility to use a "english" and a "german" UI-layout? -> Listing 2) could be fixed easily so.

**Comment by cyberjunk** [comment]
>> At first glance the static strings in the UI mentioned under 2) could be edited in this file:
"meridian59-dotnet/Resources/ui/layouts/Meridian59.layout"
Is there a possibility to use a "english" and a "german" UI-layout? -> Listing 2) could be fixed easily so.

Yes, many of the hardcoded strings (like window titles) are stored directly in the file you mentioned. But a few others are also in code files (like Constans.h in OgreClient or even in some C# corelib classes).

However what's missing is a proper way to switch the language. Anything misses here from the Options UI to the configuration value up to switching all static UI labels and:

For 1) it's the "ServerString.cs" file in Meridian59.Data.Models which needs an implementation for the parameter reordering in the BuildString() method.

**Comment by cyberjunk** [comment]
As discussed in IRC, I suggest 3 steps (3 pullrequests) for this:

STEP 1)
Add user-selected language value to the configuration.xml file and to the Options UI.
To do so at least modify the following files:
a) In C# corelib: Meridian59.Common.Config.cs (Add property/field of type Meridian59.Enums.LanguageCode, read and write its value in the XML serializers/deserializers, keep it to English as default please).
b) In OgreClient (UIOptions.cpp/UIController.h and Meridian59.layout): Add a combobox somewhere in the Options UI to let user select a language (set its value from config class instance, write back to config property - there should be plenty of examples in UIOptions to read and set config values)

At this point users can select a language without any effect...

STEP 2)
Propose a way to switch the hardcoded/static strings in the UI using the new config value. Something like a "ApplyLanguage(LanguageCode)" in the ControllerUI.h which then calls "ApplyLanguage()" on all "UIxy.cpp" submodules, which then for example sets the language dependent labels/descriptions/tooltips on each ui-module based on the current value of the Config property.

STEP 3)
Use the user selected language to build language dependent ServerString instances (this will turn the server sent chat into german). However this is not an easy task. Some parts are already done, but some important are still missing for the german message building ($ reordering)

**Comment by Ketzer2002** [comment]
Step 1 is ready to be reviewed Clint. :-)

**Comment by cyberjunk** [comment]
Please have a look at those two commits. I've built the STEP 1 based on the discussion and your pullrequests:

1) Add language value to base config class:
https://github.com/cyberjunk/meridian59-dotnet/commit/755cf73f82d47d1865a23ea0fedc059745cc713c

2) Add UI combobox to Options - Gameplay
https://github.com/cyberjunk/meridian59-dotnet/commit/0ef7bf6775b74aa3140864eb0419980f70f5fbbe

**Comment by cyberjunk** [comment]
For the STEP 2 we now need access to all the labels / descriptions that should change from code side. Most of them cannot be accessed from the code so far, because I didn't need to change them.

Here's an example which adds references for two labels in the Login window:
https://github.com/cyberjunk/meridian59-dotnet/commit/266b939de6a0e2ccf269312f4f69f7c2edf33568

**Comment by cyberjunk** [comment]
This is the commit which adds the code skeleton to switch the static UI language at runtime:
https://github.com/cyberjunk/meridian59-dotnet/commit/1386e08acdb9c23229865f9a33d73bfa2a7b2b20

The main controller ControllerUI and all sub controls have an "ApplyLanguage()" method now. Mostly empty stubs - except for the example in the UILogin.cpp which already switches the Login widget.

**Comment by cyberjunk** [comment]
I have implemented some parts of STEP 2.

Please check these files:
https://github.com/cyberjunk/meridian59-dotnet/blob/master/Meridian59.Ogre.Client/Language.h
https://github.com/cyberjunk/meridian59-dotnet/blob/master/Meridian59.Ogre.Client/Language.cpp

Technically a single big string array would have been OK, however I thought it's a better idea to group them a bit into several arrays (otherwise editing them would become a huge mess at some point..)

I could really need an helping hand now to implement all the empty ApplyLanguage() stubs.
I think the latest commits on the master give some good examples how to go on from here on.
Basically the mentioned two files above need to be extended and the ApplyLanguage() method in all the UIxy.cpp files needs an implementation.
